Source/Core/DolphinQt/Config/Graphics/BalloonTip.cpp

```diff
@@ -0,0 +1,299 @@
+bool BalloonTip::isBalloonVisible()
+{
+  return theSolitaryBalloonTip;
+}
+
+BalloonTip::BalloonTip(const QIcon& icon, const QString& title, const QString& message,
+                       QColor background, QColor text)
+    : QWidget(nullptr, Qt::ToolTip), timerId(-1), showArrow(true)
+{
+  setWindowFlags(Qt::FramelessWindowHint);
+  setAttribute(Qt::WA_NoSystemBackground);
+  setAttribute(Qt::WA_TranslucentBackground);
+  setAttribute(Qt::WA_TransparentForMouseEvents);
+  setAttribute(Qt::WA_DeleteOnClose);
+
+#if QT_CONFIG(label)
+  QLabel* titleLabel = new QLabel;
+  titleLabel->installEventFilter(this);
+  titleLabel->setText(title);
+  QFont f = titleLabel->font();
+  f.setBold(true);
+  titleLabel->setFont(f);
+  titleLabel->setTextFormat(Qt::RichText);
+#endif
+
+  const int iconSize = 18;
+
+#if QT_CONFIG(label)
+  QLabel* msgLabel = new QLabel;
+  msgLabel->installEventFilter(this);
+  msgLabel->setText(message);
+  msgLabel->setTextFormat(Qt::RichText);
+  msgLabel->setAlignment(Qt::AlignTop | Qt::AlignLeft);
+
+#if QT_VERSION < QT_VERSION_CHECK(5, 14, 0)
+  const int limit = QApplication::desktop()->availableGeometry(msgLabel).width() / 3;
+#else
+  const int limit = msgLabel->screen()->availableGeometry().width() / 3;
+#endif
+  if (msgLabel->sizeHint().width() > limit)
+  {
+    msgLabel->setWordWrap(true);
+
+    // TODO: if we are still greater than the limit we need to set the
+    // 'wrap anywhere' logic...
+    /*if (msgLabel->sizeHint().width() > limit) {
+        msgLabel->d_func()->ensureTextControl();
+        if (QWidgetTextControl *control = msgLabel->d_func()->control) {
+            QTextOption opt = control->document()->defaultTextOption();
+            opt.setWrapMode(QTextOption::WrapAnywhere);
+            control->document()->setDefaultTextOption(opt);
+        }
+    }*/
+  }
+#endif
+
+  QGridLayout* layout = new QGridLayout;
+#if QT_CONFIG(label)
+  if (!icon.isNull())
+  {
+    QLabel* iconLabel = new QLabel;
+    iconLabel->setPixmap(icon.pixmap(iconSize, iconSize));
+    iconLabel->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
+    iconLabel->setMargin(2);
+    layout->addWidget(iconLabel, 0, 0);
+    layout->addWidget(titleLabel, 0, 1);
+  }
+  else
+  {
+    layout->addWidget(titleLabel, 0, 0, 1, 2);
+  }
+#endif
+
+#if QT_CONFIG(label)
+  layout->addWidget(msgLabel, 1, 0, 1, 3);
+#endif
+  layout->setSizeConstraint(QLayout::SetFixedSize);
+  layout->setContentsMargins(10, 10, 10, 10);
+  setLayout(layout);
+
+  QPalette pal = palette();
+  pal.setColor(QPalette::Window, background);
+  pal.setColor(QPalette::WindowText, text);
+  setPalette(pal);
+}
+
+BalloonTip::~BalloonTip()
+{
+  theSolitaryBalloonTip = nullptr;
+}
+
+void BalloonTip::paintEvent(QPaintEvent*)
+{
+  QPainter painter(this);
+  painter.drawPixmap(rect(), pixmap);
+}
+
+void BalloonTip::resizeEvent(QResizeEvent* ev)
+{
+  QWidget::resizeEvent(ev);
+}
+
+void BalloonTip::balloon(const QPoint& pos, int msecs, bool showTheArrow)
```

Comment:
```suggestion
void BalloonTip::balloon(const QPoint& pos, int msecs, bool show_the_arrow)
```

msecs can likely also be a std::chrono type.